topdown/topdown_test.go

```diff
@@ -2611,6 +2654,21 @@ negation_invalidate[x] { data.a[_] = x; not data.ex.input_eq with input.x as x }
 	assertTopDownWithPath(t, compiler, store, "with vars", []string{"test", "vars"}, "", `{"foo": "hello", "bar": "world"}`)
 	assertTopDownWithPath(t, compiler, store, "with conflict", []string{"test", "conflict"}, "", fmt.Errorf("conflicting input documents"))
 	assertTopDownWithPath(t, compiler, store, "With invalidate", []string{"test", "negation_invalidate"}, "", "[2,3,4]")
+	assertTopDownWithPath(t, compiler, store, "with basic data", []string{"test", "basic_data"}, "", "true")
+	assertTopDownWithPath(t, compiler, store, "with map data overwrite", []string{"test", "map_data_1"}, "", "true")
+	assertTopDownWithPath(t, compiler, store, "with map data new key", []string{"test", "map_data_2"}, "", "true")
+	assertTopDownWithPath(t, compiler, store, "with data conflict", []string{"test", "data_conflict"}, "", fmt.Errorf("eval_with_merge_error"))
+	assertTopDownWithPath(t, compiler, store, "with base doc exact value", []string{"test", "base_doc_exact_value"}, "", `["c", "e"]`)
+	assertTopDownWithPath(t, compiler, store, "with base doc any index", []string{"test", "base_doc_any_index"}, "", `["c", "d", "e"]`)
+	assertTopDownWithPath(t, compiler, store, "undefined", []string{"test", "undefined"}, "", "")
```

Comment:
I'm not sure this test case is working as expected. For instance:

```
(detached*)torin:~/go/src/github.com/open-policy-agent/opa$ cat ~/data.json
{
    "a": {
        "b": {
            "c": [
                1,
                2,
                3
            ]
        }
    }
}(detached*)torin:~/go/src/github.com/open-policy-agent/opa$ ./opa_darwin_amd64 version
Version: 0.9.3-dev
Build Commit: deb3b985
Build Timestamp: 2018-10-16T14:59:00Z
Build Hostname: spacebox.local
(detached*)torin:~/go/src/github.com/open-policy-agent/opa$ ./opa_darwin_amd64 eval -d ~/data.json 'data'
{
  "result": [
    {
      "expressions": [
        {
          "value": {
            "a": {
              "b": {
                "c": [
                  1,
                  2,
                  3
                ]
              }
            }
          },
          "text": "data",
          "location": {
            "row": 1,
            "col": 1
          }
        }
      ]
    }
  ]
}
(detached*)torin:~/go/src/github.com/open-policy-agent/opa$ ./opa_darwin_amd64 eval -d ~/data.json 'data.a.b.c'
{
  "result": [
    {
      "expressions": [
        {
          "value": [
            1,
            2,
            3
          ],
          "text": "data.a.b.c",
          "location": {
            "row": 1,
            "col": 1
          }
        }
      ]
    }
  ]
}
(detached*)torin:~/go/src/github.com/open-policy-agent/opa$ ./opa_darwin_amd64 eval -d ~/data.json 'data.a.b.c with data.a.b as 1'
{
  "result": [
    {
      "expressions": [
        {
          "value": [
            1,
            2,
            3
          ],
          "text": "data.a.b.c",
          "location": {
            "row": 1,
            "col": 1
          }
        }
      ]
    }
  ]
}
```

Reply:
Fixed and added test case.